speaker/bgp_controller.go

```diff
@@ -272,20 +447,301 @@ type session interface {
 
 func (c *bgpController) SetNode(l log.Logger, node *v1.Node) error {
+	nodeAnnotations := node.Annotations
+	if nodeAnnotations == nil {
+		nodeAnnotations = map[string]string{}
+	}
 	nodeLabels := node.Labels
 	if nodeLabels == nil {
 		nodeLabels = map[string]string{}
 	}
-	ns := labels.Set(nodeLabels)
-	if c.nodeLabels != nil && labels.Equals(c.nodeLabels, ns) {
-		// Node labels unchanged, no action required.
+
+	anns := labels.Set(nodeAnnotations)
+	ls := labels.Set(nodeLabels)
+	annotationsUnchanged := c.nodeAnnotations != nil && labels.Equals(c.nodeAnnotations, anns)
+	labelsUnchanged := c.nodeLabels != nil && labels.Equals(c.nodeLabels, ls)
+	if labelsUnchanged && annotationsUnchanged {
+		// Node labels and annotations unchanged, no action required.
 		return nil
 	}
-	c.nodeLabels = ns
-	l.Log("event", "nodeLabelsChanged", "msg", "Node labels changed, resyncing BGP peers")
+	c.nodeAnnotations = anns
+	c.nodeLabels = ls
+
+	c.discoverNodePeer(l)
+
+	l.Log("event", "nodeChanged", "msg", "node changed, resyncing BGP peers")
 	return c.syncPeers(l)
 }
 
-var newBGP = func(logger log.Logger, addr string, myASN uint32, routerID net.IP, asn uint32, hold time.Duration, password string, myNode string) (session, error) {
-	return bgp.New(logger, addr, myASN, routerID, asn, hold, password, myNode)
+func (c *bgpController) StatusHandler() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+
+		// Copy peers slice. We want to redact BGP passwords without modifying
+		// the actual peers.
+		peers := []peer{}
+		for _, p := range c.peers {
+			cp := peer{}
+
+			cp.BGP = p.BGP
+			cfgCopy := *p.Cfg
+			cp.Cfg = &cfgCopy
+
+			cp.Cfg.Password = "REDACTED"
+
+			peers = append(peers, cp)
+		}
+
+		var np peer
+		if c.nodePeer != nil {
+			np = peer{}
+			np.BGP = c.nodePeer.BGP
+			cfgCopy := *c.nodePeer.Cfg
+			np.Cfg = &cfgCopy
+
+			np.Cfg.Password = "REDACTED"
+		}
+
+		res := struct {
+			Peers    []peer
+			NodePeer peer
+		}{
+			Peers:    peers,
+			NodePeer: np,
+		}
+
+		j, err := json.MarshalIndent(res, "", "  ")
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Failed to get status: %s", err), 500)
+			return
+		}
+		fmt.Fprint(w, string(j))
+	}
+}
+
+// parseNodePeer attempts to construct a BGP peer from information conveyed
+// in node annotations and labels using the specified autodiscovery
+// configuration.
+func parseNodePeer(l log.Logger, pad *config.PeerAutodiscovery, anns labels.Set, ls labels.Set) (*peer, error) {
+	var (
+		myASN       uint32
+		peerASN     uint32
+		peerAddr    net.IP
+		peerPort    uint16
+		srcAddr     net.IP
+		holdTime    time.Duration
+		holdTimeRaw string
+		routerID    net.IP
+	)
+
+	// Hardcoded defaults. Used only if a parameter isn't specified in the peer
+	// autodiscovery defaults and also not in annotations/labels.
+	peerPort = 179
+	holdTime = 90 * time.Second
```

Comment:
This default is in the code in two different places.  Maybe put it in a constant somewhere?

Reply:
Done (both the hold time and the default peer port).